pkg/devcontainer/config/userenvprobe.go

```diff
@@ -0,0 +1,181 @@
+package config
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/loft-sh/devpod/pkg/shell"
+	"github.com/loft-sh/log"
+)
+
+type UserEnvProbe string
+
+const (
+	LoginInteractiveShellProbe UserEnvProbe = "loginInteractiveShell"
+	LoginShellProbe            UserEnvProbe = "loginShell"
+	InteractiveShellProbe      UserEnvProbe = "interactiveShell"
+	NoneProbe                  UserEnvProbe = "none"
+
+	DefaultUserEnvProbe UserEnvProbe = LoginInteractiveShellProbe
+)
+
+func NewUserEnvProbe(probe string) (UserEnvProbe, error) {
+	switch probe {
+	case string(LoginInteractiveShellProbe):
+		return LoginInteractiveShellProbe, nil
+	case string(LoginShellProbe):
+		return LoginShellProbe, nil
+	case string(InteractiveShellProbe):
+		return InteractiveShellProbe, nil
+	case string(NoneProbe):
+		return NoneProbe, nil
+	case "":
+		return DefaultUserEnvProbe, nil
+	default:
+		return "", fmt.Errorf("invalid userEnvProbe \"%s\", supported are \"%s\"", probe,
+			strings.Join([]string{
+				string(LoginInteractiveShellProbe),
+				string(LoginShellProbe),
+				string(InteractiveShellProbe),
+				string(NoneProbe),
+			}, ","))
+	}
+}
+
+func ProbeUserEnv(ctx context.Context, probe string, userName string, log log.Logger) (map[string]string, error) {
+	userEnvProbe, err := NewUserEnvProbe(probe)
+	if err != nil {
+		log.Warnf("Get user env probe: %v", err)
+		log.Warnf("Falling back to default user env probe: %s", DefaultUserEnvProbe)
+		userEnvProbe = DefaultUserEnvProbe
+	}
+	if userEnvProbe == NoneProbe {
+		return map[string]string{}, nil
+	}
+
+	preferredShell, err := shell.GetShell(userName)
+	if err != nil {
+		return nil, fmt.Errorf("find shell for user %s: %w", userName, err)
+	}
+
+	log.Debugf("running user env probe with shell \"%s\", probe \"%s\", user \"%s\" and command \"%s\"",
+		strings.Join(preferredShell, " "), string(userEnvProbe), userName, "cat /proc/self/environ")
+
+	probedEnv, err := doProbe(ctx, userEnvProbe, preferredShell, userName, "cat /proc/self/environ", '\x00', log)
+	if err != nil {
+		log.Debugf("running user env probe with shell \"%s\", probe \"%s\", user \"%s\" and command \"%s\"",
+			strings.Join(preferredShell, " "), string(userEnvProbe), userName, "printenv")
+
+		newProbedEnv, newErr := doProbe(ctx, userEnvProbe, preferredShell, userName, "printenv", '\n', log)
+		if newErr != nil {
+			log.Warnf("failed to probe user environment variables: %v, %v", err, newErr)
+		} else {
+			probedEnv = newProbedEnv
+		}
+	}
+	if probedEnv == nil {
+		probedEnv = map[string]string{}
+	}
+
+	return probedEnv, nil
+}
+
+func getShellArgs(userEnvProbe UserEnvProbe, user, command string) []string {
```

Comment:
minor, just my preference, feel free to ignore: could we move this below `doProbe` func? This way file is a bit easier to read top to bottom,`doProbe` is first time used in `ProbeUserEnv` and `getShellArgs` is first used in `doProbe`.

Reply:
fair enough, done 👍